docs/plutus-report/scripting.tex

```diff
@@ -111,7 +115,127 @@ \subsection{Recursive values}
+  It makes much more sense to delegate certain computations to an existing tool where such computations already exist in optimized form, with good tests etc.
+\item
+  In particular, We certainly don't want to implement our own version of arithmetic (integers, addition, multiplication, etc.) or any other low-level stuff where it will be especially hard to beat external implementations.
+\end{enumerate}
+
+We certainly need efficient primitives in \gls{plutus-core}, including arithmetic.
+
+To implement this, we could build \code{Integer} into the grammar of \gls{plutus-core} types, and \code{Integer} constants into the grammar of \gls{plutus-core} terms, as well as a few functions (addition, multiplication, etc.).
+If we need more types or functions, then we simply have to extend the hardcoded list of types and functions.
+
+However, this is very inflexible.
+The key issue is that we may want to have \emph{different} sets of builtins in different situations.
+For example:
+\begin{enumerate}
+\item
+  During testing it can be convenient to have a builtin which implements \emph{tracing}: emitting log lines to a global trace output.
+  But we don't want this functionality on-chain, so we want to remove the builtin (or replace it with a no-op).
+\item
+  We may need different builtins for different ledgers (see \cref{req:script-lang-extensibility}).
+\end{enumerate}
+
+So we need to make builtin types and functions extensible.
+
+\subsubsection{Builtin types}
+
+Extensible builtin types are relatively straightforward on the type level, but become trickier on the value level.
+Values of builtin types are \emph{constants}, but what is a constant?
+It's easy to say what a \code{Integer} constant should be: an integer!
+However, what about constants of extensible builtin types?
+
+We adopt an approach familiar from generic programming in dependently-typed languages, where we parameterize our language over a \emph{universe} of types.\footnote{
+  Where do these types themselves come from?
+  We assume that we are working in some ``metalangauge'' which has types such as \code{Integer} that we can reuse in Plutus Core.
+}
+A constant then is then \emph{tagged} with a value corresponding to a type in that universe, and a value of that type.
+
+This does complicate interpretation of \gls{plutus-core} programs, since in order to interpret a program, you also need to know what universe of types it is expecting.
+In practice this is not so bad: for most users (e.g. the \gls{cardano} ledger) there will be a single universe that they use, and from there on they can simply act as though we did not have extensible builtins.
+
+\paragraph{Builtin datatypes}
+
+Adding new types to \gls{plutus-core} is not very useful unless they have functions that can operate on them!
+A particularly common case is where the new type is (morally) a datatype, and should support pattern matching.
```

Comment:
Why the "morally"?

Reply:
Hm, I guess I wanted to avoid confusion with e.g. datatypes in PIR, but I guess those aren't "in scope" here, so it's probably not confusing.

Reply:
Yeah, "morally" reads like those builtins are somehow encoded or something.